config/src/main/java/com/alibaba/nacos/config/server/service/repository/extrnal/ExternalStoragePersistServiceImpl.java

```diff
@@ -375,41 +384,53 @@ public boolean updateConfigInfo4TagCas(ConfigInfo configInfo, String tag, String
             final String srcUser, final Timestamp time, final boolean notify) {
         try {
+            int updateCount = updateConfigInfo4Beta(configInfo, betaIps, srcIp, null, time, notify);
+            if (updateCount <= 0) {
+                addConfigInfo4Beta(configInfo, betaIps, srcIp, null, time, notify);
+            }
+        } catch (EmptyResultDataAccessException ex) { // No data, insert
             addConfigInfo4Beta(configInfo, betaIps, srcIp, null, time, notify);
-        } catch (DataIntegrityViolationException ive) { // Unique constraint conflict
-            updateConfigInfo4Beta(configInfo, betaIps, srcIp, null, time, notify);
         }
     }
     
     @Override
     public boolean insertOrUpdateBetaCas(final ConfigInfo configInfo, final String betaIps, final String srcIp,
             final String srcUser, final Timestamp time, final boolean notify) {
         try {
-            addConfigInfo4Beta(configInfo, betaIps, srcIp, null, time, notify);
+            boolean updateResult = updateConfigInfo4BetaCas(configInfo, betaIps, srcIp, null, time, notify);
+            if (!updateResult) {
+                return addConfigInfo4Beta(configInfo, betaIps, srcIp, null, time, notify);
+            }
             return true;
-        } catch (DataIntegrityViolationException ive) { // Unique constraint conflict
-            return updateConfigInfo4BetaCas(configInfo, betaIps, srcIp, null, time, notify);
+        } catch (EmptyResultDataAccessException ex) { // No data, insert
+            return addConfigInfo4Beta(configInfo, betaIps, srcIp, null, time, notify);
         }
     }
     
     @Override
     public void insertOrUpdateTag(final ConfigInfo configInfo, final String tag, final String srcIp,
             final String srcUser, final Timestamp time, final boolean notify) {
         try {
+            int updateCount = updateConfigInfo4Tag(configInfo, tag, srcIp, null, time, notify);
+            if (updateCount <= 0) {
+                addConfigInfo4Tag(configInfo, tag, srcIp, null, time, notify);
+            }
+        } catch (EmptyResultDataAccessException ex) { // No data, insert
             addConfigInfo4Tag(configInfo, tag, srcIp, null, time, notify);
-        } catch (DataIntegrityViolationException ive) { // Unique constraint conflict
-            updateConfigInfo4Tag(configInfo, tag, srcIp, null, time, notify);
         }
     }
     
     @Override
     public boolean insertOrUpdateTagCas(final ConfigInfo configInfo, final String tag, final String srcIp,
             final String srcUser, final Timestamp time, final boolean notify) {
         try {
-            addConfigInfo4Tag(configInfo, tag, srcIp, null, time, notify);
+            boolean updateResult = updateConfigInfo4TagCas(configInfo, tag, srcIp, null, time, notify);
+            if (!updateResult) {
+                return addConfigInfo4Tag(configInfo, tag, srcIp, null, time, notify);
```

Comment:
if updateConfigInfo4TagCas fail，it may be cas check fail ,and should not execute add,same with above.

Reply:
Rollback cas, keep the original logic